File: lib/haveapi/action_state.rb
module HaveAPI
  # This class is an interface between APIs and HaveAPI for handling of blocking actions.
  # Blocking actions are not executed immediately, but their execution takes an unspecified
  # amount of time. This interface allows to list actions that are pending completion and view
  # their status.
  class ActionState
    # Return an array of objects representing actions that are pending completion.
    # @param [Object] user
    # @param [Integer] offset
    # @param [Integer] limit
    # @return [Array<ActionState>]
    def self.list_pending(user, offset, limit)
      raise NotImplementedError
    end

    # The constructor either gets parameter `id` or `state`. If `state` is not provided,
    # the method should find it using the `id`.
    #
    # When the client is asking about the state of a specific action, lookup using `id`
    # is used. When the client is listing pending actions, instances of this class are
    # created in self.list_pending and are passed the `state` parameter to avoid double
    # lookups. `id` should lead to the same object that would be passed as `state`.
    #
    # @param [Object] user
    # @param [Integer] id action state id
    # @param [Object] state
    def initialize(user, id: nil, state: nil)
      raise NotImplementedError
    end

    # @return [Boolean] true if the action exists
    def valid?
      raise NotImplementedError
    end

    # @return [Boolean] true of the action is finished
    def finished?
      raise NotImplementedError
    end

    # @return [Boolean] true if the action was/is going to be successful
    def status
      raise NotImplementedError
    end

    # @return [Integer] action state id
    def id
      raise NotImplementedError
    end

    # @return [String] human-readable label of this action state
    def label
      raise NotImplementedError
    end

    # @return [Hash]
    def progress
      raise NotImplementedError
    end

    # @return [Boolean] true if the action can be cancelled
    def can_cancel?
      false
    end

    # Stop action execution
    # @raise [RuntimeError] if the cancellation failed
    # @raise [NotImplementedError] if the cancellation is not supported
    # @return [Integer] if the cancellation succeded and is a blocking action
    # @return [truthy] if the cancellation succeeded
    # @return [falsy] if the cancellation failed
    def cancel
      raise NotImplementedError, 'action cancellation is not implemented by this API'
    end
  end
end

File: src/haveapi/action_state.js
/**
 * @class ActionState
 * @memberof HaveAPI.Client
 */
function ActionState (state) {
	/**
	 * @member {String} HaveAPI.Client.ActionState#label
	 * @readonly
	 */
	this.label = state.label;

	/**
	 * @member {Boolean} HaveAPI.Client.ActionState#finished
	 * @readonly
	 */
	this.finished = state.finished;

	/**
	 * @member {Boolean} HaveAPI.Client.ActionState#status
	 * @readonly
	 */
	this.status = state.status;

	/**
	 * @member {Boolean} HaveAPI.Client.ActionState#canCancel
	 * @readonly
	 */
	this.canCancel = state.can_cancel;

	/**
	 * @member {HaveAPI.Client.ActionState.Progress} HaveAPI.Client.ActionState#progress
	 * @readonly
	 */
	this.progress = new ActionState.Progress(state);
};

/**
 * Stop tracking of this action state
 * @method HaveAPI.Client.ActionState#stop
 */
ActionState.prototype.stop = function () {
	this.doStop = true;
};

/**
 * @method HaveAPI.Client.ActionState#shouldStop
 */
ActionState.prototype.shouldStop = function () {
	return this.doStop || false;
};

/**
 * @method HaveAPI.Client.ActionState#cancel
 */
ActionState.prototype.cancel = function () {
	// TODO
};

/**
 * @class HaveAPI.Client.ActionState.Progress
 * @memberof HaveAPI.Client.ActionState
 */
ActionState.Progress = function (state) {
	/**
   * @member {Integer} HaveAPI.Client.ActionState.Progress.current
	 * @readonly
	 */
	this.current = state.current;

	/**
	 * @member {Integer} HaveAPI.Client.ActionState.Progress.total
	 * @readonly
	 */
	this.total = state.total;

	/**
	 * @member {String} HaveAPI.Client.ActionState.Progress.unit
	 * @readonly
	 */
	this.unit = state.unit;
};

/**
 * @method HaveAPI.Client.ActionState.Progress#toString
 * @return {String}
 */
ActionState.Progress.prototype.toString = function () {
	return this.current + "/" + this.total + " " + this.unit;
};
